rust/src/index/prefilter.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Secondary Index pre-filter
+//!
+//! Based on the query, we might have information about which fragment ids and
+//! row ids can be excluded from the search.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use futures::stream::BoxStream;
+use futures::StreamExt;
+
+use crate::error::Result;
+use crate::Dataset;
+
+/// Filter out row ids that we know are not relevant to the query. This currently
+/// is just deleted rows.
```

Comment:
FYI in the future I hope to re-use this struct to also do fragment-level pruning based on statistics.